File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/TransferInfo.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * Return info about a transfer/initiate transfer (source->target)
 * Possibly should have a Params field like above
 *
 * <p>Template: {@code TransferInfo Low 154 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code processTransferInfo()} in indra/llmessage/lltransfermanager.cpp
 * (secondlife/viewer @ c179f76c01).
 */
public class TransferInfo extends SLMessage {
    public TransferInfoData TransferInfoData_Field;

    public static class TransferInfoData {
        public int ChannelType; // S32
        public byte[] Params; // Variable 2
        public int Size; // S32
        public int Status; // S32
        public int TargetType; // S32
        public UUID TransferID; // LLUUID
    }

    public TransferInfo() {
        this.zeroCoded = true;
        this.TransferInfoData_Field = new TransferInfoData();
    }

    @Override
    public int CalcPayloadSize() {
        return this.TransferInfoData_Field.Params.length + 34 + 4;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleTransferInfo(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 154 (TransferInfo).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x9A);
        packUUID(byteBuffer, this.TransferInfoData_Field.TransferID);
        packInt(byteBuffer, this.TransferInfoData_Field.ChannelType);
        packInt(byteBuffer, this.TransferInfoData_Field.TargetType);
        packInt(byteBuffer, this.TransferInfoData_Field.Status);
        packInt(byteBuffer, this.TransferInfoData_Field.Size);
        packVariable(byteBuffer, this.TransferInfoData_Field.Params, 2);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.TransferInfoData_Field.TransferID = unpackUUID(byteBuffer);
        this.TransferInfoData_Field.ChannelType = unpackInt(byteBuffer);
        this.TransferInfoData_Field.TargetType = unpackInt(byteBuffer);
        this.TransferInfoData_Field.Status = unpackInt(byteBuffer);
        this.TransferInfoData_Field.Size = unpackInt(byteBuffer);
        this.TransferInfoData_Field.Params = unpackVariable(byteBuffer, 2);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/TransferInfo.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * Return info about a transfer/initiate transfer (source->target)
 * Possibly should have a Params field like above
 *
 * <p>Template: {@code TransferInfo Low 154 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code processTransferInfo()} in indra/llmessage/lltransfermanager.cpp
 * (secondlife/viewer @ c179f76c01).
 */
open class TransferInfo : SLMessage() {
    @JvmField var TransferInfoData_Field: TransferInfoData = TransferInfoData()

    open class TransferInfoData {
        @JvmField var ChannelType: Int = 0
        @JvmField var Params: ByteArray? = null
        @JvmField var Size: Int = 0
        @JvmField var Status: Int = 0
        @JvmField var TargetType: Int = 0
        @JvmField var TransferID: UUID? = null
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return TransferInfoData_Field.Params!!.size + 34 + 4
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleTransferInfo(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 154 (TransferInfo).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x9A).toByte())
        packUUID(byteBuffer, TransferInfoData_Field.TransferID)
        packInt(byteBuffer, TransferInfoData_Field.ChannelType)
        packInt(byteBuffer, TransferInfoData_Field.TargetType)
        packInt(byteBuffer, TransferInfoData_Field.Status)
        packInt(byteBuffer, TransferInfoData_Field.Size)
        packVariable(byteBuffer, TransferInfoData_Field.Params, 2)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        TransferInfoData_Field.TransferID = unpackUUID(byteBuffer)
        TransferInfoData_Field.ChannelType = unpackInt(byteBuffer)
        TransferInfoData_Field.TargetType = unpackInt(byteBuffer)
        TransferInfoData_Field.Status = unpackInt(byteBuffer)
        TransferInfoData_Field.Size = unpackInt(byteBuffer)
        TransferInfoData_Field.Params = unpackVariable(byteBuffer, 2)
    }
}
